src/execution_plan/ops/op_foreach.c

```diff
@@ -99,30 +94,49 @@ static Record ForeachConsume
 	op->first = false;
 
 	// construct an array of records to hold all consumed records (eagerly)
+	// and an array holding clones of the input records with the mapping of the
+	// embedded plan
 	op->records = array_new(Record, 1);
+	op->body_records = array_new(Record, 1);
+
+	// insert a NULL value to op->records, so that execution terminates when it
+	// is consumed by the parent
+	array_append(op->records, NULL);
 
 	Record r = NULL;
 	if(op->supplier) {
 		// eagerly drain supplier
 		while((r = OpBase_Consume(op->supplier))) {
 			array_append(op->records, r);
+
+			// create a record with the mapping of the embedded plan
+			// (as opposed to the record-mapping of the consumed record)
+			Record body_rec = OpBase_CreateRecord(op->body);
+			// copy the consumed record's entries to the record to be sent to
+			// the body
+			Record_DeepClone(r, body_rec);
```

Comment:
`DeepClone` is a bit of an overkill, we don't want to clone the actual `SIValues`, we can pass them as volatile or we can transfer the ownership.